dev/com.ibm.ws.security.authorization.jacc.ejb/src/com/ibm/ws/security/authorization/jacc/ejb/impl/EJBSecurityValidatorImpl.java

```diff
@@ -84,66 +90,110 @@ private boolean checkMethodConstraints(final String contextId,
-                                PolicyContext.setContextID(contextId);
+                                               new PrivilegedExceptionAction<Boolean>() {
+                                                   @Override
+                                                   public Boolean run() throws javax.security.jacc.PolicyContextException {
+                                                       PolicyContext.setContextID(contextId);
 
-                                if (tc.isDebugEnabled())
-                                    Tr.debug(tc, "Registering JACC context handlers");
-                                for (String key : jaccHandlerKeyArray) {
-                                    PolicyContext.registerHandler(key, pch, true);
-                                }
-                                handlerObjects.put(jaccHandlerKeyArray[0], subject);
-                                handlerObjects.put(jaccHandlerKeyArray[1], bean);
-                                handlerObjects.put(jaccHandlerKeyArray[2], methodParameters);
-                                MessageContext mc = null;
-                                try {
-                                    InitialContext ic = new InitialContext();
-                                    mc = getMessageContext(ic);
-                                } catch (NamingException e) {
-                                    if (tc.isDebugEnabled())
-                                        Tr.debug(tc, "NamingException is caught. Ignoring..", e);
-                                }
-                                if (mc != null) {
-                                    if (tc.isDebugEnabled())
-                                        Tr.debug(tc, "javax.xml.soap.SOAPMessage is set: ", mc);
-                                    handlerObjects.put(jaccHandlerKeyArray[3], mc);
-                                }
-                                ProtectionDomain pd = null;
-
-                                if (subject != null && subject.getPrincipals().size() > 0) {
-                                    Principal[] principalArray = subject.getPrincipals().toArray(new Principal[subject.getPrincipals().size()]);
-                                    pd = new ProtectionDomain(nullCs, null, null, principalArray);
-                                } else {
-                                    pd = nullPd;
-                                }
+                                                       if (tc.isDebugEnabled())
+                                                           Tr.debug(tc, "Registering JACC context handlers");
+                                                       for (String key : jaccHandlerKeyArray) {
+                                                           PolicyContext.registerHandler(key, pch, true);
+                                                       }
+                                                       handlerObjects.put(jaccHandlerKeyArray[0], subject);
```

Comment:
I am concerned about the jaccHandlerKeyArray.  Does there need to be jakarta version of those properties instead of the java EE ones.  You can use something like what is done here to help with some of this:

https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ScheduledTask.java#L431

You can have an eeVersion put into your Component to allow you to know if you are doing EE9 or less than EE9.  You can see how to do that in the concurrent project.  If you need help, let me know.  But you can avoid the calling of getMessageContext(InitialContext) and some of the follow on code if it is EE9 or above.

You could also consider to have a second impl of this class where you call a method to do the MessageContext and it does nothing for EE9, but does the old way for EE8 and less.

Reply:
I have made the suggested changes by injecting the JavaEEVersion service into the EJBServiceImpl and setting the EE version in EJBSecurityValidatorImpl.

Reply:
You did not address whether you need a second version of the jaccHandlerKeyArray for jakarta.  Have the property names changed in Jakarta EE 9.  Maybe that is something forth coming.  The link I pointed to shows how they handle the different properties depending on version and reading them to support both the old javax... and jakarta... property names.

Reply:
It appears the only key that changed is `javax.ejb.arguments` as it changed to `jakarta.ejb.arguments`. I will update the code here.

See https://jakarta.ee/specifications/authorization/2.0/authorization-spec-2.0-SNAPSHOT.html for details.

Reply:
I expect that this will change and it will be jakarta.ejb.EnterpriseBean because the spec hasn't been updated yet to remove the javax reference.  Something to follow up with the spec people.

Reply:
When ejbCore-2.0 and this function is actually used it will be interesting what the test cases end up doing.

Reply:
I have opened an issue for the spec here:

https://github.com/eclipse-ee4j/authorization/issues/80

I have also made created (#12955) a blocking issue for this PR's parent issue (#11857) to make any changes as a result of work done on the above issue.

I have updated EJBSecurityValidatorImpl.java to set both the ee8- and ee9 keys appropriately. I have also updated PolicyContextHandlerImpl to return the super-set of both the old and the new properties, while retaining the original order. These changes will potentially make it easier for customer to migrate to ee9.